### ISSUE TITLE
dense-matrix mul!(C, A, B, alpha, beta) allocates

### ISSUE THREAD

**Issue by amilsted** [JuliaLang/julia#46865]
On Julia 1.8.1 I noticed the following:
```julia
function manymul(N, C, A, B, alpha, beta)
    for i in 1:N
        mul!(C, A, B, alpha, beta)
        #BLAS.gemm!('N', 'N', alpha, A, B, beta, C)  # eliminates allocations
        C, A = A, C
    end
    C
end

D = 16
A = randn(D, D)
B = randn(D, D)
C = zero(A)

N = 100000
@time manymul(N, C, A, B, 1.0, 0.5)   #allocates N times (32 bytes each) with `mul!()`, 0 times with `gemm!()`
```
Cthulhu suggests this is due to runtime dispatch related to `MulAdd()`. This can impact performance of e.g. ODE solving involving `mul!()` for small matrix sizes. The example above takes around 10% longer with `mul!()` vs. `gemm!()`, according to benchmarktools (single-threaded BLAS).

Is this known/intended?

My `versioninfo()`:
```
Julia Version 1.8.1
Commit afb6c60d69a (2022-09-06 15:09 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin21.4.0)
  CPU: 12 × Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, skylake)
  Threads: 1 on 6 virtual cores
Environment:
  JULIA_EDITOR = code
  JULIA_NUM_THREADS = 
  JULIA_PKG_USE_CLI_GIT = true
```

**Comment by giordano** [comment]
Isn't that because you're accessing non-`const` globals? Does it change if you make those arrays `const`? (I'm away from computer now, can't try myself)

**Comment by PallHaraldsson** [comment]
No, it allocates 3*N times even if I put into a function. I don't know if this is intended, nor if promised not to allocate. I did comment out e.g. this line that allocates extra: C, A = A, C

**Comment by PallHaraldsson** [comment]
It's strange, it's defined like:
```
@inline function mul!(C::StridedMatrix{T}, A::StridedVecOrMat{T}, B::StridedVecOrMat{T},
                             alpha::Number, beta::Number) where {T<:BlasFloat}
           return gemm_wrapper!(C, 'N', 'N', A, B, MulAddMul(alpha, beta))
       end
```

but I ruled out gemm_wrapper! and MulAddMul problematic (but maybe not BlasFloat, I cloned mul! to mul2! to test but substituted with Float64):

```
@time gemm_wrapper!(C, 'N', 'N', A, B, MulAddMul(1.0, 1.5))
  0.000015 seconds
```

**Comment by PallHaraldsson** [comment]
I did (do still?) suspect threading, because I actually DID get 3*N allocations (on 1.7.3, and likely remembering right for 1.8.1), but then that went away down to 0 [N] allocations, and then I get N allocations.

I'm pasting verbatim, to show I'm not wrong (and the function definitions should be dummy redefinitions):
```
julia> @time manymul(N, C, A, B, 1.0, 0.5)
  0.000024 seconds (3 allocations: 64 bytes)

julia> N
1

julia> function manymul(N, C, A, B, alpha, beta)
                         for i in 1:N
                             #mul!(C, A, B, alpha, beta)
                             BLAS.gemm!('N', 'N', alpha, A, B, beta, C)  # eliminates allocations
                             # C, A = A, C
                         end
                         nothing # C
                     end
manymul (generic function with 1 method)

julia> @time manymul(N, C, A, B, 1.0, 0.5)
  0.018076 seconds (14.48 k allocations: 848.309 KiB, 99.68% compilation time)

julia> @time manymul(N, C, A, B, 1.0, 0.5)
  0.000022 seconds

julia> function manymul(N, C, A, B, alpha, beta)
                         for i in 1:N
                             mul!(C, A, B, alpha, beta)
                             #BLAS.gemm!('N', 'N', alpha, A, B, beta, C)  # eliminates allocations
                             # C, A = A, C
                         end
                         nothing # C
                     end
manymul (generic function with 1 method)

julia> @time manymul(N, C, A, B, 1.0, 0.5)
  0.032734 seconds (14.71 k allocations: 852.665 KiB, 99.85% compilation time)

julia> @time manymul(N, C, A, B, 1.0, 0.5)
  0.000021 seconds (1 allocation: 32 bytes)
```

I did run again with 1.7.3 with:
```
$ julia -t1
```

and got 3 allocations. Maybe I ruled out regular threads, but not BLAS threads.

**Comment by PallHaraldsson** [comment]
running on 1.8.1 (trying to rule out BLAS threads, but I'm not sure OMP_NUM_THREADS does anything in my version):
```
julia> @time for i in 1:1000 manymul(N, C, A, B, 1.0, 0.5) end
  0.001940 seconds (1000 allocations: 31.250 KiB)
```

running the same in 1.7.3 I got:
```
julia> @time for i in 1:1000 manymul(N, C, A, B, 1.0, 0.5) end
  0.002373 seconds (3.00 k allocations: 62.500 KiB)
```

Three times the allocations (likely not always 3*N as I wrote about before), but not 3*n bytes, if that's helpful for anyone to know to anyone, only 2*N allocations.

**Comment by PallHaraldsson** [comment]
I think I rule out (BLAS) threads an issue below:
```
julia> versioninfo()
Julia Version 1.8.1
Commit afb6c60d69a (2022-09-06 15:09 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 16 × Intel(R) Xeon(R) CPU D-1541 @ 2.10GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, broadwell)
  Threads: 1 on 16 virtual cores

julia> @time @time @time mul!(C, A, B, 1.0, 1.5)
  0.000016 seconds (1 allocation: 32 bytes)
  0.004511 seconds (45 allocations: 2.570 KiB, 97.40% compilation time)
  0.004548 seconds (72 allocations: 4.961 KiB, 96.60% compilation time)

julia> BLAS.set_num_threads(1)  # still allocates after this, and this should be effective I can send in (0) or even (-1) though...

julia> @time @time @time mul!(C, A, B, 1.0, 1.5)
  0.000017 seconds (1 allocation: 32 bytes)
  0.000110 seconds (26 allocations: 1.477 KiB)
  0.000135 seconds (50 allocations: 3.148 KiB)
```

In that terminal I was however runing julia, not julia -t1

**Comment by amilsted** [comment]
I added version info to the report above. On my system the allocation behavior (post compilation) seems to be deterministic and it must be something in `mul!()` that is not in `gemm!()`, as the `gemm!()` version does not allocate. `@benchmark` also agrees that the `mul!()` version allocates, whereas the `gemm!()` version does not.

**Comment by amilsted** [comment]
This version also does *not* allocate:
```
function manymul(N, out, op, in, alpha, beta)
    _add = LinearAlgebra.MulAddMul{false, false, typeof(alpha), typeof(beta)}(alpha, beta)
    for _ in 1:N
        LinearAlgebra.gemm_wrapper!(out, 'N', 'N', op, in, _add)
        out, in = in, out
    end
    out
end
```
The difference is that a fully-concrete `MulAddMul` is used.

**Comment by giordano** [comment]
Now that I'm back to the computer, I'm more and more convinced the only thing here is that you're accessing non-`const` globals:
```julia
julia> using LinearAlgebra

julia> function manymul(N, C, A, B, alpha, beta)
           for i in 1:N
               mul!(C, A, B, alpha, beta)
               C, A = A, C
           end
           C
       end
manymul (generic function with 1 method)

julia> const D = 16;

julia> const A = randn(D, D);

julia> const B = randn(D, D);

julia> const C = zero(A);

julia> const N = 100000;

julia> @time manymul(N, C, A, B, 1.0, 0.5);
  0.075428 seconds
```
So I don't see what's the issue here.

**Comment by PallHaraldsson** [comment]
Yes, cons fixed for me too, but shouldn't this work do, while it doesnt?
```
function test()
       D = 16
       A = randn(D, D)
       B = randn(D, D)
       C = zero(A)

       N = 100000
       @time manymul(N, C, A, B, 1.0, 0.5)
end

julia> @time test()
  0.117295 seconds (100.00 k allocations: 3.052 MiB)
  0.117421 seconds (100.03 k allocations: 3.060 MiB)
```

**Comment by amilsted** [comment]
Interesting! I also see 0 allocations in @giordano's example. However:
```julia
D = 16
const A = randn(D, D)
const B = randn(D, D)
const C = zero(A)
const N = 100000
const alphas = [1.0]

@time manymul(N, C, A, B, alphas[1], 0.5)   #allocates N times
```
Considering also @PallHaraldsson's latest evidence, it seems we rely on constant propagation of alpha and beta to avoid allocation, **and it is quite fragile even when alpha and beta are constants**.

Personally, I don't think we should be relying on constant prop. here, at least not in the BLAS case - I don't think reading alpha or beta from a vector is a particularly crazy thing to do - and also the constant cases seem to be fragile in ways I wouldn't have expected. @PallHaraldsson's example is particularly odd!

**Comment by PallHaraldsson** [comment]
>we rely on constant propagation of alpha and beta to avoid allocation, **and it is quite fragile**.

You mean non-deterministic. I would have thought the compiler compiles the same way each time for same inputs, is that not to be relied upon, and why might that be?

**Comment by amilsted** [comment]
> You mean non-deterministic. I would have thought the compiler compiles the same way each time for same inputs, is that not to be relied upon, and why might that be?

I didn't observe any non-determinism. Constant propagation depends on the compiler figuring out that certain values are constant and can be converted to some kind of `Const` type. Functions called with different types involve different methods. I do think it is fragile, though, since it seems to be easy to prevent the compiler from doing this. In your example, I can get const. prop. to work by adding an `@inline` to the `manymul()` callsite.

**Comment by PallHaraldsson** [comment]
I mean non-deterministinc (allocations, then none, then back on) as I showed there:
https://github.com/JuliaLang/julia/issues/46865#issuecomment-1255637879

Yes, I did "compile", but same code in-between.

**Comment by amilsted** [comment]
Huh, that is indeed puzzling.

Btw, I took a peek at `matmul.jl` and friends in LinearAlgebra - `MulAdd` is everywhere! Even if constant prop were working perfectly, why would we only want non-allocating versions of all these functions for constant alpha and beta? If I am using a 5-arg `mul!` I am trying to squeeze out as much performance as I can! I agree with @Micket's comments on slack: seems like a questionable choice.

At least in `gemm_wrapper!`, it looks like one could quite easily push `MulAdd` down the call stack so that it is not needed for the BLAS case at least (where it really is entirely superfluous).

**Comment by amilsted** [comment]
Just want to add that these allocations, albeit small, can be disastrous for scaling with threaded parallelism due to GC contention. And yes, we have seen this in real code.

**Comment by amilsted** [comment]
Oh man, this issue has history. https://github.com/JuliaLang/julia/issues/34013, https://github.com/JuliaLang/julia/pull/29634

**Comment by amilsted** [comment]
Worth noting that the 2x2 and 3x3 versions of the tests above *also* allocate.

**Comment by amilsted** [comment]
I wrote a macro that pulls the value-dependent branching for `MulAddMul()` outside the enclosing function call. I'm sure this is not the prettiest way to do it:
```julia
macro mambranch(expr)
    expr.head == :call || throw(ArgumentError("Can only handle function calls."))
    for (i, e) in enumerate(expr.args)
        e isa Expr || continue
        if e.head == :call && e.args[1] == :(LinearAlgebra.MulAddMul)
            local asym = e.args[2]
            local bsym = e.args[3]

            local e_sub11 = copy(expr)
            e_sub11.args[i] = :(LinearAlgebra.MulAddMul{true, true, typeof($asym), typeof($bsym)}($asym, $bsym))
            
            local e_sub10 = copy(expr)
            e_sub10.args[i] = :(LinearAlgebra.MulAddMul{true, false, typeof($asym), typeof($bsym)}($asym, $bsym))
            
            local e_sub01 = copy(expr)
            e_sub01.args[i] = :(LinearAlgebra.MulAddMul{false, true, typeof($asym), typeof($bsym)}($asym, $bsym))

            local e_sub00 = copy(expr)
            e_sub00.args[i] = :(LinearAlgebra.MulAddMul{false, false, typeof($asym), typeof($bsym)}($asym, $bsym))

            local e_out = quote
                if isone($asym) && iszero($bsym)
                    $e_sub11
                elseif isone($asym)
                    $e_sub10
                elseif iszero($bsym)
                    $e_sub01
                else
                    $e_sub00
                end
            end
            return esc(e_out)
        end
    end
    throw(ArgumentError("No valid MulAddMul expression found."))
end
```

With the macro, you can write a fully inferable `manymul` like this:
```julia
function manymul(N, out, op, in, alpha, beta)
    for _ in 1:N
        @mambranch LinearAlgebra.gemm_wrapper!(out, 'N', 'N', op, in, LinearAlgebra.MulAddMul(alpha, beta))
        out, in = in, out
    end
    out
end
```
It gets transformed by the macro into approximately this:
```julia
function manymul(N, out, op, in, alpha, beta)
    for _ in 1:N
        ais1, bis0 = isone(alpha), iszero(beta)
        if ais1 && bis0
            LinearAlgebra.gemm_wrapper!(out, 'N', 'N', op, in, LinearAlgebra.MulAddMul{true, true, typeof(alpha), typeof(beta)}(alpha, beta))
        elseif ais1
            LinearAlgebra.gemm_wrapper!(out, 'N', 'N', op, in, LinearAlgebra.MulAddMul{true, false, typeof(alpha), typeof(beta)}(alpha, beta))
        elseif bis0
            LinearAlgebra.gemm_wrapper!(out, 'N', 'N', op, in, LinearAlgebra.MulAddMul{false, true, typeof(alpha), typeof(beta)}(alpha, beta))
        else
            LinearAlgebra.gemm_wrapper!(out, 'N', 'N', op, in, LinearAlgebra.MulAddMul{false, false, typeof(alpha), typeof(beta)}(alpha, beta))
        end
        out, in = in, out
    end
    out
end
```

This does not allocate for either const or variable alpha/beta! Adding something like this to `mul!()` in `matmul.jl` seems a fair way of removing the const-prop fragility and make non-const cases also nonallocating.

I don't see any disadvantages: This branching won't happen at runtime in the const-prop case, and was happening in any case in the `MulAddMul()` constructor otherwise. Now it just doesn't lead to runtime dispatch.

**Comment by amilsted** [comment]
There ought to be a simpler solution that exploits union-splitting?

**Comment by jlchan** [comment]
Any update on this issue?

**Comment by amilsted** [comment]
Should be fixed in 1.12! https://github.com/JuliaLang/julia/pull/52439
